# Extending Logistic Regression: 2009 American Community Survey

### Nick Chao

### Preparation and Overview (20 points total)

[10 points] Explain the task and what business-case or use-case it is designed to solve (or designed to investigate). Detail exactly what the classification task is and what parties would be interested in the results.

[5 points] (mostly the same processes as from previous labs) Define and prepare your class variables. Use proper variable representations (int, float, one-hot, etc.). Use pre-processing methods (as needed) for dimensionality reduction, scaling, etc. Remove variables that are not needed/useful for the analysis. Describe the final dataset that is used for classification/regression (include a description of any newly formed variables you created).

[5 points] Divide you data into training and testing data using an 80% training and 20% testing split. Use the cross validation modules that are part of scikit-learn. Argue "for" or "against" splitting your data using an 80/20 split. That is, why is the 80/20 split appropriate (or not) for your dataset? 


### Modeling (50 points total)

[20 points] Create a custom, one-versus-all logistic regression classifier using numpy and scipy to optimize. Use object oriented conventions identical to scikit-learn. You should start with the template developed by the instructor in the course. You should add the following functionality to the logistic regression classifier:
Ability to choose optimization technique when class is instantiated: either steepest descent, stochastic gradient descent, or Newton's method. 
Update the gradient calculation to include a customizable regularization term (either using no regularization, L1 regularization, L2 regularization, or both L1 and L2 regularization). Associate a cost with the regularization term, "C", that can be adjusted when the class is instantiated.  

[15 points] Train your classifier to achieve good generalization performance. That is, adjust the optimization technique and the value of the regularization term "C" to achieve the best performance on your test set. Visualize the performance of the classifier versus the parameters you investigated. Is your method of selecting parameters justified? That is, do you think there is any "data snooping" involved with this method of selecting parameters?

[15 points] Compare the performance of your "best" logistic regression optimization procedure to the procedure used in scikit-learn. Visualize the performance differences in terms of training time and classification performance. Discuss the results. 

### Deployment (10 points total)
Which implementation of logistic regression would you advise be used in a deployed machine learning model, your implementation or scikit-learn (or other third party)? Why?

### Exceptional Work (10 points total)
You have free reign to provide additional analyses. One idea: Update the code to use either "one-versus-all" or "one-versus-one" extensions of binary to multi-class classification. 
One idea (required for 7000 level students): Implement an optimization technique for logistic regression using mean square error as your objective function (instead of binary entropy). Your solution should be able to solve the binary logistic regression problem in one gradient update step. 


## Data Overview

In this lab, we investigate possible relationships between an individual's income and attributes about them. The data used in this lab is provided by the 2009 American Community Survey 1-Year PUMS Population File which can be found here. https://catalog.data.gov/dataset/2009-american-community-survey-1-year-pums-population-file. This dataset contains more than 3 million entries and nearly 300 attributes. To make sense of some of these attributes, there is a dictionary lookup that provides more information about the columns. You can find this reference here. https://www2.census.gov/programs-surveys/acs/tech_docs/pums/data_dict/PUMSDataDict09.pdf.

To be more specific, we want to see if we can predict a person's current income based on factors about themselves that they might give away when applying for a new job. For example, their age, sex, education level, and more are just a few peices of information that companies may ask for when applying for a new job. 

This information could be incrediably useful for a company that is hiring new personal. If they know the current income of someone who has applied for a job at their company, then they can offer the lowest starting salary they believe the new canadate will accept. (i.e. slightly above what they currently make). 


In [36]:
#importing dependancies
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import ShuffleSplit
from sklearn.linear_model import LogisticRegression


In [2]:
%time dataA = pd.read_csv('../data/ss09pusa.csv')
%time dataB = pd.read_csv('../data/ss09pusb.csv')
merged = pd.concat([dataA,dataB])

Wall time: 27.6 s
Wall time: 25.7 s


In [3]:
merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3030728 entries, 0 to 1466654
Columns: 279 entries, RT to pwgtp80
dtypes: float64(86), int64(190), object(3)
memory usage: 6.3+ GB


In [4]:
print('The 2009 American Community Survey has a lot of data, look at all these columns! \n\n'+str(list(merged.columns.values)))

The 2009 American Community Survey has a lot of data, look at all these columns! 

['RT', 'SERIALNO', 'SPORDER', 'PUMA', 'ST', 'ADJINC', 'PWGTP', 'AGEP', 'CIT', 'CITWP', 'COW', 'DDRS', 'DEAR', 'DEYE', 'DOUT', 'DPHY', 'DRAT', 'DRATX', 'DREM', 'ENG', 'FER', 'GCL', 'GCM', 'GCR', 'HINS1', 'HINS2', 'HINS3', 'HINS4', 'HINS5', 'HINS6', 'HINS7', 'INTP', 'JWMNP', 'JWRIP', 'JWTR', 'LANX', 'MAR', 'MARHD', 'MARHM', 'MARHT', 'MARHW', 'MARHYP', 'MIG', 'MIL', 'MLPA', 'MLPB', 'MLPC', 'MLPD', 'MLPE', 'MLPF', 'MLPG', 'MLPH', 'MLPI', 'MLPJ', 'MLPK', 'NWAB', 'NWAV', 'NWLA', 'NWLK', 'NWRE', 'OIP', 'PAP', 'REL', 'RETP', 'SCH', 'SCHG', 'SCHL', 'SEMP', 'SEX', 'SSIP', 'SSP', 'WAGP', 'WKHP', 'WKL', 'WKW', 'WRK', 'YOEP', 'ANC', 'ANC1P', 'ANC2P', 'DECADE', 'DIS', 'DRIVESP', 'ESP', 'ESR', 'FOD1P', 'FOD2P', 'HICOV', 'HISP', 'INDP', 'JWAP', 'JWDP', 'LANP', 'MIGPUMA', 'MIGSP', 'MSP', 'NAICSP', 'NATIVITY', 'NOP', 'OC', 'OCCP', 'PAOC', 'PERNP', 'PINCP', 'POBP', 'POVPIP', 'POWPUMA', 'POWSP', 'PRIVCOV', 'PUBCOV', 'QTRBIR

### Data Preparation

Obviously there is way more data here than we need so lets start by reducing the number of columns to only what we consider useful for our classification. The following attributes will remain as they are peices of information that new hires might give away when applying for a job.

Citizenship, age, class of work, English fluentcy, martial status, military status, sex, education background, disability status, race, geographical location. We will also keep the individual's income as it is the attribute that we are attempting to predict. 

In [19]:
cols_to_save = ['CIT','AGEP','COW','ENG','MAR','MIL','SCHL','SEX','DIS','PINCP','POWSP','RAC1P','FOD1P']
new_data = merged.filter(items=cols_to_save)
new_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3030728 entries, 0 to 1466654
Data columns (total 13 columns):
CIT      int64
AGEP     int64
COW      float64
ENG      float64
MAR      int64
MIL      float64
SCHL     float64
SEX      int64
DIS      int64
PINCP    float64
POWSP    float64
RAC1P    int64
FOD1P    float64
dtypes: float64(7), int64(6)
memory usage: 323.7 MB


In [20]:
new_data

CIT  AGEP  COW  ENG  MAR  MIL  SCHL  SEX  DIS     PINCP  POWSP  \
0          1    51  NaN  NaN    4  3.0  16.0    1    1    3800.0    NaN   
1          1    64  1.0  NaN    1  5.0  19.0    2    1   36800.0    1.0   
2          1    68  1.0  NaN    1  3.0  14.0    1    1   54600.0    1.0   
3          1    61  1.0  NaN    3  5.0  16.0    2    2    6000.0    1.0   
4          1    38  1.0  NaN    5  5.0  16.0    1    2   14000.0    1.0   
5          1    65  6.0  NaN    1  5.0  16.0    2    1   13000.0    NaN   
6          1    74  1.0  NaN    1  3.0  16.0    1    1   45200.0    1.0   
7          1    23  1.0  NaN    5  5.0  12.0    2    1     820.0    NaN   
8          1    42  1.0  NaN    1  5.0  19.0    2    2   25200.0    1.0   
9          1    42  1.0  NaN    1  5.0  13.0    1    2   56000.0    1.0   
10         1    23  1.0  NaN    3  5.0  18.0    2    2       0.0    NaN   
11         1     3  NaN  NaN    5  NaN   2.0    2    2       NaN    NaN   
12         1    35  3.0  NaN    3  5.0  19.0    1    2   45000.0    1.0   
13         1    29  1.0  NaN    3  5.0  19.0    2    2     300.0    NaN   
14         1    39  1.0  NaN    1  5.0  16.0    1    2   41000.0    1.0   
15         1    38  1.0  NaN    1  5.0  16.0    2    2    3500.0    1.0   
16         1    14  NaN  NaN    5  NaN  11.0    1    2       NaN    NaN   
17         1    13  NaN  NaN    5  NaN   9.0    2    2       NaN    NaN   
18         1    12  NaN  NaN    5  NaN   8.0    2    2       NaN    NaN   
19         1    32  NaN  NaN    5  5.0  12.0    1    2    1500.0    NaN   
20         1    34  4.0  NaN    5  5.0  16.0    2    2   14200.0    1.0   
21         1    14  NaN  NaN    5  NaN  10.0    2    2       NaN    NaN   
22         1    12  NaN  NaN    5  NaN   8.0    1    2       NaN    NaN   
23         1    72  NaN  NaN    1  3.0  17.0    1    1   12000.0    NaN   
24         1    60  1.0  NaN    1  5.0  16.0    2    2       0.0    NaN   
25         1    39  7.0  NaN    3  5.0  19.0    1    1   20500.0    1.0   
26         1    53  1.0  NaN    1  5.0  21.0    2    2   66000.0    1.0   
27         1    54  1.0  NaN    1  5.0  19.0    1    2   45000.0    1.0   
28         1    21  1.0  NaN    5  5.0  16.0    1    2    6000.0    NaN   
29         1    51  2.0  NaN    5  5.0  24.0    1    2   97520.0    1.0   
...      ...   ...  ...  ...  ...  ...   ...  ...  ...       ...    ...   
1466625    1    75  NaN  NaN    2  5.0  16.0    2    2   37500.0    NaN   
1466626    5    30  1.0  2.0    1  5.0   9.0    1    2   50000.0   56.0   
1466627    5    26  NaN  2.0    1  5.0  15.0    2    2       0.0    NaN   
1466628    1     4  NaN  NaN    5  NaN   1.0    1    2       NaN    NaN   
1466629    1     2  NaN  NaN    5  NaN   NaN    2    2       NaN    NaN   
1466630    1    48  1.0  NaN    1  5.0  21.0    1    2   80000.0   56.0   
1466631    1    53  1.0  NaN    1  5.0  19.0    2    2    3000.0   56.0   
1466632    1    25  1.0  NaN    5  5.0  18.0    1    2    3700.0   30.0   
1466633    1    93  NaN  NaN    2  5.0  16.0    2    1   12000.0    NaN   
1466634    1    64  4.0  NaN    2  5.0  16.0    2    1   10800.0    NaN   
1466635    1    81  NaN  NaN    1  3.0  19.0    1    2   28700.0    NaN   
1466636    1    78  NaN  NaN    1  5.0  16.0    2    2    6800.0    NaN   
1466637    1    79  NaN  NaN    2  3.0  21.0    1    2   53400.0    NaN   
1466638    1    87  NaN  NaN    2  5.0  12.0    2    1    4200.0    NaN   
1466639    1    79  NaN  NaN    2  5.0  18.0    2    2   41100.0    NaN   
1466640    1    49  2.0  NaN    1  5.0  18.0    2    2   46000.0   56.0   
1466641    1    55  1.0  NaN    1  5.0  16.0    1    2   22000.0   56.0   
1466642    1    24  3.0  NaN    1  5.0  21.0    2    2   13020.0   56.0   
1466643    1     1  NaN  NaN    5  NaN   NaN    2    2       NaN    NaN   
1466644    1    29  3.0  NaN    1  5.0  21.0    1    2   36500.0   56.0   
1466645    1    56  NaN  NaN    3  5.0  18.0    2    2    6000.0    NaN   
1466646    1    42  1.0  NaN    3 

We began with over 3 million records and 279 differnt attributes provided in the survey. Of the nearly 300 attributes that were provided in the dataset, we reduced it down to the thirteen shown above. The table below shows the desired attributes and examples of what we would want the data to look like. 

|Attribute|Description|Type|Example|
|:---:|:---:|:---:|:---:|
| CIT | Citizenship Status | Bool | 1. Citizen, 0. Non-citizen |
| AGEP | Age | Int | 23
| COW | Class of Worker | Float | 3. Local Government, 4. State Government |
| ENG | Ability to speak English  | Bool | 1. Speaks English, 0. Doesn't Speak English |
| MAR | Marital Status | Int | 1. Married, 2. Widowed |
| MIL | Military Service | Bool | 1. Yes, 0. No |
| SCHL | Educational Attainment  | Float | 21 Bachelor's Degree, 22 Master's Degree |
| SEX | Sex      |Bool | True. Male |
| DIS | Disability | Bool | True. Disabled |
| PINCP | Total Person's Income | Float
| POWSP | Place of work | Float | 048 Texas, 049 Utah |
| RAC1P | Detailed Race Code | Int | 1 White, 6 Asian |
| FOD1P | Field of Degree | Float | 2407 Computer Engineering, 2408 Electrical Engineering |

### Data Cleaning

The next step is data cleaning and modifying the data into more useful data types. Lets start by modifying some of these data types into more useful and proper variables.

In [21]:
# Change citizenship to boolean.
# 1-4 is a citizen (true) and 5 is not a citizen (false)

new_data.CIT.replace(to_replace = range(5),
                    value=[1,1,1,1,0],
                    inplace=True)
new_data['CIT'] = new_data['CIT'].astype('bool')

In [22]:
# Change Ability to Speak English to boolean
# b is N/A but it would be a good assumption to assume they speak English
new_data['ENG']=new_data['ENG'].fillna(1)
# 1-2 speaks English well or very well, 3-4 speaks English not well or not at all.
new_data.ENG.replace(to_replace = range(4),
                    value=[1,1,0,0],
                    inplace=True)
new_data['ENG'] = new_data['ENG'].astype('bool')# Change Military Status to Boolean

In [23]:
# b is N/A because less than 17 years old so lets just change this to 0
new_data['MIL']=new_data['MIL'].fillna(0)
# 1-3 Yes, 4-5 No
new_data.MIL.replace(to_replace = range(5),
                    value=[1,1,1,0,0],
                    inplace=True)
new_data['MIL'] = new_data['MIL'].astype('bool')

In [24]:
# Change Sex to Boolean 
# 1 is male, 2 is female. Changing 2 to 0 for boolean conversion
new_data.SEX.replace(to_replace = range(2),
                    value=[1,0],
                    inplace=True)
new_data['SEX'] = new_data['SEX'].astype('bool')

In [25]:
# Change DIS to Boolean 
# 1 is disabled, 2 is no disability. Changing 2 to 0 for boolean conversion
new_data.DIS.replace(to_replace = range(2),
                    value=[1,0],
                    inplace=True)
new_data['DIS'] = new_data['DIS'].astype('bool')

In [26]:
# Change Educational Atttainment to INT
# bb is N/A for less than 3 years old.
new_data['SCHL']=new_data['SCHL'].fillna(0)
# For this classification lets simplify some of these education levels.
# 0 between No schooling and Grade 8
# 1 between Grade 9 and Grade 12 no diploma
# 2 for High School degree or GED
# 3 Some college to Associate's degree
# 4 Bachelor's Degree
# 5 Master's Degree
# 6 Professional degree or Doctorate
new_data.SCHL.replace(to_replace = range(25),
                    value=[0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,2,2,3,3,3,4,5,6,6],
                    inplace=True)
new_data['SCHL'] = new_data['SCHL'].astype('int')

Next, lets remove all entries with people under the age of 18 because our goal is to focus on personal income of working class individuals.

In [27]:
# delete younger than 18
new_data = new_data[new_data.AGEP >= 18]
#new_data

Now lets work on the Nulls...

In [28]:
# find null columns
print('Columns that contain nulls: '+str(new_data.columns[new_data.isnull().any()].tolist()))

Columns that contain nulls: ['COW', 'POWSP', 'FOD1P']


For Field of Study, we can replace all the Nulls with 0s since they only refer to those with at least a college degree
Let's remove any Class of Worker and Place of Work rows with Nulls since those entres are for idividuals who have not worked

In [29]:
# Field of Study  -> 0
# Class of Worker -> Remove if Null
# Place of Work   -> Remove if Null

new_data['FOD1P'].fillna(0, inplace=True)
new_data = new_data[pd.notnull(new_data['COW'])]
new_data = new_data[pd.notnull(new_data['POWSP'])]

C:\Users\Nick\Anaconda3\lib\site-packages\pandas\core\generic.py:4355: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)


In [30]:
# Double check for null columns
print('Columns that contain nulls: '+str(new_data.columns[new_data.isnull().any()].tolist()))

Columns that contain nulls: []


Let's split the personal income into 5 different classes. This will help verify our income predictions while also being specific enough to work with our business goal.

In [31]:
new_data['PINCP'] = pd.qcut(new_data.PINCP, 5, labels=[0,1,2,3,4])
new_data['PINCP'].unique()

[2, 3, 0, 1, 4]
Categories (5, int64): [0 < 1 < 2 < 3 < 4]

In [32]:
new_data['PINCP'] = new_data['PINCP'].astype(np.int)

In [33]:
# Lets see how the income classes have split...
print('Number of people in each class:')
for value in new_data.PINCP.unique(): 
    print(str(value)+': ' +str(len(new_data[new_data['PINCP'] == value])))

Number of people in each class:
2: 233964
3: 268048
0: 267705
1: 302608
4: 265954


In [34]:
# Finally, let's rename some of these columns so they make more sense.
new_data.rename(columns={'CIT': 'Citizenship','AGEP': 'Age','COW': 'Class of Work','ENG': 'Speaks English','MAR': 'Martial Status','MIL': 'Military Status','SCHL': 'Education Level','SEX': 'Male','DIS': 'Disabled?','PINCP': 'Income','POWSP': 'Place of Work','RAC1P': 'Race','FOD1P': 'Field of Study'}, inplace=True)
new_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1338279 entries, 1 to 1466654
Data columns (total 13 columns):
Citizenship        1338279 non-null bool
Age                1338279 non-null int64
Class of Work      1338279 non-null float64
Speaks English     1338279 non-null bool
Martial Status     1338279 non-null int64
Military Status    1338279 non-null bool
Education Level    1338279 non-null int32
Male               1338279 non-null bool
Disabled?          1338279 non-null bool
Income             1338279 non-null int32
Place of Work      1338279 non-null float64
Race               1338279 non-null int64
Field of Study     1338279 non-null float64
dtypes: bool(5), float64(3), int32(2), int64(3)
memory usage: 88.1 MB


### Training and Testing

As we can see above, there is still a significant amount of data for testing and training purposes (Over 1.3 million entries). The next step is to determine how much of the data should be used immediately for training purposes and how much should be saved until the end for testing purposes. 

An 80/20 split seems to be a good number for this situation. As long as the split is randomized and there would be over a million entries just for training, and around a quarter of a million entries for testing. That seems like plenty of data to create a accurate model as well as test the model for it's accuracy. Furthermore, since we split the income into five classes, the 80/20 split seems appropriate.

In [40]:
# Lets start by making a copy of the cleaned data we are using.
new_df = new_data.copy()
if 'Income' in new_data:
    y = new_df['Income'].values    # Since Income is our target class, lets make a copy of it
    del new_df['Income']           # Now we need to remove the target class
    X = new_df.values              # The remaining data will be used to train

# Scikit Learns provides a way to split our data into training and testing subsets.
cv_object = ShuffleSplit(train_size=.8, test_size=0.2, n_splits=1)
                         
print(cv_object)

ShuffleSplit(n_splits=1, random_state=None, test_size=0.2, train_size=0.8)


## Modeling